pkg/services/ngalert/store/alert_rule.go

```diff
@@ -507,6 +519,15 @@ func (st DBstore) UpdateRuleGroup(cmd UpdateRuleGroupCmd) error {
 			return err
 		}
 
+		// delete instances for rules that will not be removed
+		for _, rule := range existingGroupRules {
+			if _, ok := existingGroupRulesUIDs[rule.UID]; !ok {
+				if err := st.DeleteAlertInstanceByRuleUID(cmd.OrgID, rule.UID); err != nil {
+					return err
+				}
+			}
+		}
```

Comment:
Why not just loop on `existingGroupRulesUIDs` like below?

Reply:
Because what we need the inverse of existingGroupRulesUIDs. There are two things to do: 

- delete the rule, which also deletes instances, for rules that existed before the upsert but are not part of the upsert
- delete only the instances for the rules that were removed from existingGroupRulesUIDs